### PR TITLE
Do not parse content tags to generate excerpt

### PR DESCRIPTION
**Pull request by felixarntz**
This PR temporarily unhooks `wp_filter_content_tags()` from `the_content` filter when invoking it in `wp_trim_excerpt()`.
* Since that function only gets the content to generate the excerpt, which never includes any HTML tags, calling `wp_filter_content_tags()` is wasteful and unnecessary in that context.
* More importantly, calling it there leads to bugs like the one outlined in the Trac ticket.

Trac ticket: https://core.trac.wordpress.org/ticket/56588

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

### CODE REVIEW

File: tests/phpunit/tests/formatting/wpTrimExcerpt.php [hunk 1]
```diff
@@ -92,4 +92,49 @@ public function test_should_generate_excerpt_for_empty_values() {
...
+			}
+		);
+
+		wp_trim_excerpt( '', $post );
+
+		// Assert that the filter callback was unhooked while running 'the_content'.
+		$this->assertFalse( $has_filter, 'wp_filter_content_tags() was not unhooked in wp_trim_excerpt()' );
+
+		// Assert that the filter callback was restored after running 'the_content'.
+		$this->assertSame( 10, has_filter( 'the_content', 'wp_filter_content_tags' ), 'wp_filter_content_tags() was not restored in wp_trim_excerpt()' );
```
Comment (kt-12) [inline tests/phpunit/tests/formatting/wpTrimExcerpt.php:132]: In light of what was discussed [here](https://github.com/WordPress/wordpress-develop/pull/4424#discussion_r1186750067) should we use two different functions for each assertions.

File: src/wp-includes/formatting.php [hunk 2]
```diff
@@ -3914,10 +3914,26 @@ function wp_trim_excerpt( $text = '', $post = null ) {
...
+		/*
+		 * Temporarily unhook wp_filter_content_tags() since any tags
+		 * within the excerpt are stripped out. Modifying the tags here
+		 * is wasteful and can lead to bugs in the image counting logic.
+		 */
+		$filter_removed = false;
+		if ( has_filter( 'the_content', 'wp_filter_content_tags' ) ) {
+			remove_filter( 'the_content', 'wp_filter_content_tags' );
+			$filter_removed = true;
+		}
```
Comment (spacedmonkey) [inline src/wp-includes/formatting.php]:
```suggestion
		$filter_removed = remove_filter( 'the_content', 'wp_filter_content_tags' );
```

Could we make this simpler by doing this? `remove_filter` returns a boolean.

File: tests/phpunit/tests/formatting/wpTrimExcerpt.php (hunk 1 above)
Comment (felixarntz) [reply tests/phpunit/tests/formatting/wpTrimExcerpt.php:132]: Let's wait to hear back on https://github.com/WordPress/wordpress-develop/pull/4390#discussion_r1194245533 first, as I'm not sure that's a strict requirement. Given that the assertions have distinct messages, it may be okay as is. cc @costdev

File: src/wp-includes/formatting.php (hunk 2 above)
Comment (felixarntz) [reply src/wp-includes/formatting.php]: That's a great idea, will update to simplify.

File: tests/phpunit/tests/formatting/wpTrimExcerpt.php (hunk 1 above)
Comment (costdev) [reply tests/phpunit/tests/formatting/wpTrimExcerpt.php:132]: Thanks for the ping @felixarntz!

These tests don't appear to be testing that the `loading="lazy"` is absent, which is the bug in the linked ticket. While the current tests to ensure the filter is removed/restored should definitely stay, there should also be a test to verify that this change has the desired effect.

Looking at this test method, it does appear that two different things are being tested:
1. That the content is not run through `wp_filter_content_tags()`.
2. That calling `wp_trim_excerpt()` doesn't permanently unhook `wp_filter_content_tags()` from `the_content`.

I'd suggest splitting these. For (2), it should be quite straightforward:
```php
/**
 * Tests that `wp_trim_excerpt()` doesn't permanently
 * unhook 'wp_filter_content_tags()` from `'the_content'`.
 *
 * @ticket 56588
 *
 * @covers ::wp_trim_excerpt
 */
public function test_wp_trim_excerpt_should_not_permanently_unhook_wp_filter_content_tags() {
	$post = self::factory()->post->create();

	/*
	 * Ensure that `wp_filter_content_tags()` has been hooked
	 * before testing that it hasn't been permanently removed.
	 */
	if ( false === has_filter( 'the_content', 'wp_filter_content_tags' ) ) {
		add_filter( 'the_content', 'wp_filter_content_tags' );
	}

	$this->assertSame(
		10,
		has_filter( 'the_content', 'wp_filter_content_tags' ) ,
		'The "wp_filter_content_tags" callback was not hooked to "the_content".'
	);

	wp_trim_excerpt( '', $post );

	$this->assertSame(
		10,
		has_filter( 'the_content', 'wp_filter_content_tags' ),
		'The "wp_filter_content_tags" callback was permanently unhooked.'
	);
}
```

File: src/wp-includes/formatting.php [hunk 3]
```diff
@@ -3914,10 +3914,22 @@ function wp_trim_excerpt( $text = '', $post = null ) {
 		$text = strip_shortcodes( $text );
 		$text = excerpt_remove_blocks( $text );
 
+		/*
+		 * Temporarily unhook wp_filter_content_tags() since any tags
+		 * within the excerpt are stripped out. Modifying the tags here
+		 * is wasteful and can lead to bugs in the image counting logic.
+		 */
+		$filter_removed = remove_filter( 'the_content', 'wp_filter_content_tags' );
```
Comment (costdev) [inline src/wp-includes/formatting.php:3945]: @felixarntz Is it possible that we could ever change the priority that `wp_filter_content_tags()` is hooked to `'the_content'`?

Just wondering if we should be storing the result of `has_filter()` here, and if `is_int( $has_filter )`, then remove the filter. Then when restoring, we can pass the priority back into the `add_filter()` call.

Comment (spacedmonkey) [inline src/wp-includes/formatting.php:3945]:
```suggestion
                 $priority = has_filter(  'the_content', 'wp_filter_content_tags' );
                 $filter_removed = false;
                 if(  $priority ){
		     $filter_removed = remove_filter( 'the_content', 'wp_filter_content_tags', $priority );
		}
```

@costdev Is this what you had in mind?

**Review by spacedmonkey — CHANGES_REQUESTED** [review]

File: src/wp-includes/formatting.php (hunk 3 above)
Comment (costdev) [reply src/wp-includes/formatting.php:3945]: @spacedmonkey Similar to that yeah.

```php
$filter_priority = has_filter( 'the_content', 'wp_filter_content_tags' );

if ( is_int( $filter_priority ) ) { // Or false !== $filter_priority
    remove_filter( 'the_content', 'wp_filter_content_tags' );
}
```

Later:
```php
if ( is_int( $filter_priority ) ) { // Or false !== $filter_priority
    add_filter( 'the_content', 'wp_filter_content_tags', $filter_priority );
}
```

- `has_filter()` will return `int|false` in this context, which could even be `0`, so a type check/strict check is needed.
- The call to `remove_filter()` is dependent on the result of `has_filter()`, so we don't need to store the result of `remove_filter()` - we know it'll be `true`.

**Comment by mukeshpanchal27** [comment]
@felixarntz, I tried to reproduce the issue, but I couldn't. Could you please provide the steps to reproduce the issue?

File: src/wp-includes/formatting.php (hunk 3 above)
Comment (felixarntz) [reply src/wp-includes/formatting.php:3945]: @costdev That seems unnecessary to me as WP core is in control about the filter priority so if we were to change it in the future (which is unlikely as that would be a BC break), we could update it in both places.

Yes, it makes the code a bit more flexible, but it also adds another check which IMO isn't really necessary.

Comment (felixarntz) [reply src/wp-includes/formatting.php:3945]: See my reply in https://github.com/WordPress/wordpress-develop/pull/4439#discussion_r1198081475. I'm not strongly opposed, but it seems unnecessary overhead to me.

Comment (spacedmonkey) [reply src/wp-includes/formatting.php:3945]: It feel a bit of a edge case for us to worry about. If a developer has double hooked this on, they may have a good reason.

Let's flag this in the dev note.

Comment (spacedmonkey) [reply src/wp-includes/formatting.php:3945]: +1 to @felixarntz comment. It is a worry, but it is an edge case. Something to flag in the dev note.

**Comment by felixarntz** [comment]
@mukeshpanchal27 In order to test this, you need to use a theme that on its archive pages (e.g. the blog) shows post excerpts (`the_excerpt()`) and post featured images (`the_post_thumbnail()`).

Assuming that the theme displays the featured image first, and then the excerpt, for each post, you'll need to do the following:
1. Create a post that has a featured image.
2. Create another post (i.e. newer than the first post) that doesn't have a featured image, but has 3 images within its post content.
3. View the blog, where the 2. post should show up first, and the 1. post should show up second, since it's older. The only image you should see is the featured image of the one post, since excerpts are always plain text and don't include images.
4. Without this PR, you should notice that the featured image has `loading="lazy"` on it, which is incorrect since it's the first image on the blog. With this PR, the featured image shouldn't have `loading="lazy"` anymore, which would be correct.

File: tests/phpunit/tests/formatting/wpTrimExcerpt.php (hunk 1 above)
Comment (joemcgill) [reply tests/phpunit/tests/formatting/wpTrimExcerpt.php:132]: I agree with @costdev's point about these tests not actually testing the bug this change is intended to fix. Adding a specific test case that demonstrates the bug and showing that it is fixed would be best. Ideally, asserting that the image count is not effected when `wp_trim_excerpt()` is called would be good.

File: src/wp-includes/formatting.php [hunk 4]
```diff
@@ -3914,10 +3914,22 @@ function wp_trim_excerpt( $text = '', $post = null ) {
...
+
 		/** This filter is documented in wp-includes/post-template.php */
 		$text = apply_filters( 'the_content', $text );
 		$text = str_replace( ']]>', ']]&gt;', $text );
 
+		// Only restore the filter callback if it was removed above.
+		if ( $filter_removed ) {
+			add_filter( 'the_content', 'wp_filter_content_tags' );
```
Comment (joemcgill) [inline src/wp-includes/formatting.php:3957]: Someone could have changed the priority of this filter intentionally, so we should put it back with the priority exactly how we found it.

```suggestion
			add_filter( 'the_content', 'wp_filter_content_tags', $filter_removed );
```

File: tests/phpunit/tests/formatting/wpTrimExcerpt.php [hunk 5]
```diff
@@ -92,4 +92,49 @@ public function test_should_generate_excerpt_for_empty_values() {
...
+		$post = self::factory()->post->create();
+
+		/*
+		 * Record that during 'the_content' filter run by wp_trim_excerpt() the
+		 * wp_filter_content_tags() callback is not used.
+		 */
+		$has_filter = true;
+		add_filter(
+			'the_content',
+			function( $content ) use ( &$has_filter ) {
```
Comment (mukeshpanchal27) [inline tests/phpunit/tests/formatting/wpTrimExcerpt.php]:
```suggestion
			static function( $content ) use ( &$has_filter ) {
```
Use static closures whenever $this is not used to avoid memory leaks.

File: tests/phpunit/tests/formatting/wpTrimExcerpt.php [hunk 6]
```diff
@@ -92,4 +92,49 @@ public function test_should_generate_excerpt_for_empty_values() {
 		$this->assertSame( 'Post content', wp_trim_excerpt( null, $post ) );
 		$this->assertSame( 'Post content', wp_trim_excerpt( false, $post ) );
 	}
+
+	/**
+	 * @ticket 56588
+	 * @covers ::wp_trim_excerpt
```
Comment (mukeshpanchal27) [inline tests/phpunit/tests/formatting/wpTrimExcerpt.php]:
```suggestion
```
Remove the `@covers` annotation as it is already defined in the class, so there is no need to add it to individual tests.

File: tests/phpunit/tests/formatting/wpTrimExcerpt.php [hunk 7]
```diff
@@ -92,4 +92,49 @@ public function test_should_generate_excerpt_for_empty_values() {
...
+		// Assert that the filter callback was unhooked while running 'the_content'.
+		$this->assertFalse( $has_filter, 'wp_filter_content_tags() was not unhooked in wp_trim_excerpt()' );
+
+		// Assert that the filter callback was restored after running 'the_content'.
+		$this->assertSame( 10, has_filter( 'the_content', 'wp_filter_content_tags' ), 'wp_filter_content_tags() was not restored in wp_trim_excerpt()' );
+	}
+
+	/**
+	 * @ticket 56588
+	 * @covers ::wp_trim_excerpt
```
Comment (mukeshpanchal27) [inline tests/phpunit/tests/formatting/wpTrimExcerpt.php]:
```suggestion
```

Check my comment above.

**Review by mukeshpanchal27 — COMMENTED** [review]
Thanks, @felixarntz. I have left some feedback. However, I apologize as I was unable to reproduce the same issue in Twenty Twenty-One and Twenty Nineteen.

File: src/wp-includes/formatting.php (hunk 4 above)
Comment (spacedmonkey) [reply src/wp-includes/formatting.php:3957]: `$filter_removed` is a boolean. If we want to `$priority`, we will have to do something like this. 

https://github.com/WordPress/wordpress-develop/pull/4439/files#r1194977298

Comment (felixarntz) [reply src/wp-includes/formatting.php:3957]: This was discussed here: https://github.com/WordPress/wordpress-develop/pull/4439#discussion_r1194653054

Note that `remove_filter()` doesn't return the priority, instead you have to know the priority when you call that function.

File: tests/phpunit/tests/formatting/wpTrimExcerpt.php (hunk 1 above)
Comment (felixarntz) [reply tests/phpunit/tests/formatting/wpTrimExcerpt.php:132]: @costdev @joemcgill In https://github.com/WordPress/wordpress-develop/pull/4439/commits/823fbe4fc667d6a96b9aafffc30e52adaa20b3d9, I've split the existing tests accordingly.

In https://github.com/WordPress/wordpress-develop/pull/4439/commits/d373f638ea985d07a2f383cffa1cef7e80c6d58d, I've added 2 more comprehensive test around the concrete bug fixed: One tests that the lower-level content media count is not affected, the other at a higher level that as a result of that lazy-loading is omitted correctly.

**Comment by felixarntz** [comment]
@mukeshpanchal27 
> I was unable to reproduce the same issue in Twenty Twenty-One and Twenty Nineteen.

Have you ensured you test with a theme and template that uses `the_excerpt()` and not `the_content()`?
* For Twenty Twenty-One you'll need to make sure the setting is set to display excerpts.
* For Twenty Nineteen, there is no such setting and it always uses `the_content()`, except in the search template. So in Twenty Nineteen, you would need to create 2 posts that satisfy a specific search query, and then use the search to validate that the bug is fixed.

**Comment by felixarntz** [comment]
@mukeshpanchal27 I took a closer look at the test instructions again, see https://core.trac.wordpress.org/ticket/56588#comment:20 for the main clarification.

File: tests/phpunit/tests/media.php [hunk 8]
```diff
@@ -3948,6 +3948,105 @@ public function data_special_contexts_for_the_content() {
...
+		$post_id           = self::factory()->post->create(
+			array(
+				'post_content' => $post_content,
+				'post_excerpt' => '',
+			)
+		);
+		$featured_image_id = self::$large_id;
+		update_post_meta( $post_id, '_thumbnail_id', $featured_image_id );
+
+		$expected_output  = wpautop( '<p>Some text.</p>' );
```
Comment (mukeshpanchal27) [inline tests/phpunit/tests/media.php]:
```suggestion
		$expected_output  = wpautop( 'Some text.' );
```

**Review by mukeshpanchal27 — APPROVED** [review]
Thanks, @felixarntz, for providing the steps to reproduce the issue. The PR resolves the edge case problem. I have left one non-blocker review; other than that, it looks good to me.

**Review by costdev — APPROVED** [review]
Tests well, LGTM! :+1:

**Review by spacedmonkey — APPROVED** [review]

File: src/wp-includes/formatting.php (hunk 3 above)
Comment (westonruter) [reply src/wp-includes/formatting.php:3945]: A couple cases of prior art in core:

The `wp_pre_kses_block_attributes()` function:

https://github.com/WordPress/wordpress-develop/blob/b1f2a760a283c24511623e1c13421f20e27fed88/src/wp-includes/formatting.php#L5162-L5164

The `WP_Widget_Text::widget()` method:

https://github.com/WordPress/wordpress-develop/blob/472c9f242ba6d9141bdb4d239aa9c725abb53214/src/wp-includes/widgets/class-wp-widget-text.php#L249-L253

https://github.com/WordPress/wordpress-develop/blob/472c9f242ba6d9141bdb4d239aa9c725abb53214/src/wp-includes/widgets/class-wp-widget-text.php#L321-L324

Comment (felixarntz) [reply src/wp-includes/formatting.php:3945]: @westonruter No alignment on how to handle the priority between those cases either, it seems 😆 

I personally think checking for the priority is overkill, given it is controlled by core and not intended to be changed by plugin developers. Removing one of those crucial `the_content` filters or changing their order has a high chance to break other plugins and therefore is not recommended.

File: tests/phpunit/tests/media.php [hunk 9]
```diff
@@ -3948,6 +3948,105 @@ public function data_special_contexts_for_the_content() {
...
+		$post_id = self::factory()->post->create(
+			array(
+				'post_content' => $post_content,
+				'post_excerpt' => '',
+			)
+		);
+
+		$wp_query     = new WP_Query( array( 'post__in' => array( $post_id ) ) );
+		$wp_the_query = $wp_query;
+		$this->reset_content_media_count();
```
Comment (joemcgill) [inline tests/phpunit/tests/media.php]: Does this not automatically get reset during every test method? If not, seems like we should move this to a fixture to avoid cross-contamination between tests.

File: src/wp-includes/formatting.php (hunk 4 above)
Comment (joemcgill) [reply src/wp-includes/formatting.php:3957]: Right, sorry for missing the return value difference in my previous comment. From a strict backwards compatibility perspective, I think we need to reset the filter to the same priority in which we found it since plugins could be intentionally moving the priority of this function for any reason.

File: tests/phpunit/tests/media.php [hunk 10]
```diff
@@ -3948,6 +3948,105 @@ public function data_special_contexts_for_the_content() {
 		);
 	}
 
+	/**
+	 * Tests that the content media count is not affected by `the_excerpt()` calls for posts that contain images.
+	 *
+	 * @ticket 56588
+	 *
+	 * @covers ::the_excerpt
+	 * @covers ::wp_trim_excerpt
```
Comment (joemcgill) [inline tests/phpunit/tests/media.php]: While this test makes sure these two functions aren't triggering side effects, we're not really testing those functions, but are instead testing `wp_increase_content_media_count()`,  so maybe we should update this to the following?:

```suggestion
	 * @covers :: wp_increase_content_media_count
```

File: tests/phpunit/tests/media.php [hunk 11]
```diff
@@ -3948,6 +3948,105 @@ public function data_special_contexts_for_the_content() {
...
+		);
+		$featured_image_id = self::$large_id;
+		update_post_meta( $post_id, '_thumbnail_id', $featured_image_id );
+
+		$expected_output  = wpautop( 'Some text.' );
+		$expected_output .= get_the_post_thumbnail( $post_id, 'post-thumbnail', array( 'loading' => false ) );
+
+		$wp_query     = new WP_Query( array( 'post__in' => array( $post_id ) ) );
+		$wp_the_query = $wp_query;
+		$this->reset_content_media_count();
```
Comment (joemcgill) [inline tests/phpunit/tests/media.php]: Same as above

File: tests/phpunit/tests/formatting/wpTrimExcerpt.php [hunk 12]
```diff
@@ -92,4 +92,60 @@ public function test_should_generate_excerpt_for_empty_values() {
...
+	 *
+	 * @ticket 56588
+	 */
+	public function test_wp_trim_excerpt_unhooks_wp_filter_content_tags() {
+		$post = self::factory()->post->create();
+
+		/*
+		 * Record that during 'the_content' filter run by wp_trim_excerpt() the
+		 * wp_filter_content_tags() callback is not used.
+		 */
```
Comment (westonruter) [inline tests/phpunit/tests/formatting/wpTrimExcerpt.php:107]: Maybe overkill but assert it was originally present:

```suggestion
		 */
		 $this->assertSame( 10, has_filter( 'the_content', 'wp_filter_content_tags' ) );
```

File: tests/phpunit/tests/media.php [hunk 13]
```diff
@@ -3948,6 +3948,105 @@ public function data_special_contexts_for_the_content() {
...
+		$this->reset_omit_loading_attr_filter();
+
+		$output = '';
+		while ( have_posts() ) {
+			the_post();
+
+			// Print excerpt first, then the featured image.
+			$output .= get_echo( 'the_excerpt' );
+			$output .= get_echo( 'the_post_thumbnail' );
```
Comment (joemcgill) [inline tests/phpunit/tests/media.php:4156]: Is this trying to test that the excerpt is being rendered correctly and that the post thumbnail is not being lazy loaded? If only the latter, it may be clearer to do a string check to make sure `loading="lazy"` is not present on the image.

File: tests/phpunit/tests/formatting/wpTrimExcerpt.php [hunk 14]
```diff
@@ -92,4 +92,60 @@ public function test_should_generate_excerpt_for_empty_values() {
...
+			'the_content',
+			static function( $content ) use ( &$has_filter ) {
+				$has_filter = has_filter( 'the_content', 'wp_filter_content_tags' );
+				return $content;
+			}
+		);
+
+		wp_trim_excerpt( '', $post );
+
+		$this->assertFalse( $has_filter, 'wp_filter_content_tags() was not unhooked in wp_trim_excerpt()' );
```
Comment (westonruter) [inline tests/phpunit/tests/formatting/wpTrimExcerpt.php:119]: Assert it gets restored:

```suggestion
		$this->assertSame( 10, has_filter( 'the_content', 'wp_filter_content_tags' ) );
```

**Review by joemcgill — CHANGES_REQUESTED** [review]
The new tests are looking good, thanks for adding them! I still think we need to retain the filter priority unless others feel strongly otherwise. My feedback on the tests are not blockers, just suggestions.

File: tests/phpunit/tests/formatting/wpTrimExcerpt.php (hunk 14 above)
Comment (westonruter) [reply tests/phpunit/tests/formatting/wpTrimExcerpt.php:119]: Granted, this is also done in `test_wp_trim_excerpt_should_not_permanently_unhook_wp_filter_content_tags`

File: tests/phpunit/tests/media.php (hunk 9 above)
Comment (felixarntz) [reply tests/phpunit/tests/media.php]: @joemcgill What do you mean by fixture? You mean in a `set_up()` method for example?

Indeed this isn't automatically reset since it relies on a static variable inside the function.

File: tests/phpunit/tests/media.php (hunk 13 above)
Comment (felixarntz) [reply tests/phpunit/tests/media.php:4156]: @joemcgill Fair point, though most existing tests use full content assertions, so this isn't new. It still makes the tests a bit more comprehensive, for example if we just check that the output doesn't contain `loading="lazy"`, what about if suddenly the output contains no `img` tag at all? The test would still pass, but something would be broken - admittedly probably something else, but I think given that there is no real overhead in comparing the full string and it's been done for years in such test scenarios, I'd prefer to keep it as is.

Comment (felixarntz) [reply tests/phpunit/tests/media.php:4156]: @joemcgill Fair point, though most existing tests use full content assertions, so this isn't new. It still makes the tests a bit more comprehensive, for example if we just check that the output doesn't contain `loading="lazy"`, what about if suddenly the output contains no `img` tag at all? The test would still pass, but something would be broken - admittedly probably something else, but I think given that there is no real overhead in comparing the full string and it's been done for years in such test scenarios, I'd prefer to keep it as is.

File: src/wp-includes/formatting.php (hunk 4 above)
Comment (westonruter) [reply src/wp-includes/formatting.php:3957]: @joemcgill I'm not sure this will be necessary since the the third argument to `remove_filter()` is `$priority` and has a default value of `10`. So this means that if `$filter_removed` is `true`, the filter had a priority of `10` and which is the default in `add_filter()` here too.

File: tests/phpunit/tests/media.php (hunk 10 above)
Comment (felixarntz) [reply tests/phpunit/tests/media.php]: @joemcgill Here I disagree. The `wp_increase_content_media_count()` function was not touched as part of this PR, so nothing here is testing that function. It's testing how the function is called, which is happening in `wp_trim_excerpt()`, hooked into `the_excerpt` filter.

File: src/wp-includes/formatting.php (hunk 4 above)
Comment (felixarntz) [reply src/wp-includes/formatting.php:3957]: @joemcgill I'd prefer to leave this as is, as plugins shouldn't change the priorities of `the_content` filter - it'll likely break other plugins. More importantly, the way this is currently here is not a backward compatibility break since, if a plugin has changed it, we wouldn't unhook and rehook the filter in any case, as @westonruter mentioned above. Effectively, if a plugin has changed that filter, they wouldn't get the benefits of this fix, but nothing would actually break / get worse than it was before.

But I don't feel too strongly about that, so happy to adjust.

File: tests/phpunit/tests/formatting/wpTrimExcerpt.php (hunk 12 above)
Comment (felixarntz) [reply tests/phpunit/tests/formatting/wpTrimExcerpt.php:107]: @westonruter I have been avoiding those kinds of assertions, to generally avoid multiple assertions in the same test, which is fine here since those situations are tested in another test, so we don't miss anything really.

File: tests/phpunit/tests/formatting/wpTrimExcerpt.php (hunk 14 above)
Comment (felixarntz) [reply tests/phpunit/tests/formatting/wpTrimExcerpt.php:119]: Exactly, related is my reply in https://github.com/WordPress/wordpress-develop/pull/4439#discussion_r1202696771.

File: tests/phpunit/tests/media.php (hunk 9 above)
Comment (joemcgill) [reply tests/phpunit/tests/media.php]: Yup, that what I meant! If the count is not reset before each method that tests image output, then the unit tests themselves can cause side effects to other tests. Looks like we're already doing this manually in several methods, so moving it to a shared fixture would make the whole test class a bit cleaner and more predictable.

File: tests/phpunit/tests/media.php (hunk 10 above)
Comment (joemcgill) [reply tests/phpunit/tests/media.php]: My thinking here is that `@covers` is meant to specify [which functions should be shown in coverage reports as being covered by unit tests](https://docs.phpunit.de/en/9.5/annotations.html#covers) and filter out others used in the method, not just to specify which functions are being called within the test method.

This one is a tricky one because we're trying to test that calling `wp_trim_excerpt ` in specific cases doesn't cause side effects by increasing the content media count unnecessarily, but we're not really testing the output of `the_excerpt` or `wp_trim_excerpt`. Perhaps `wp_trim_excerpt` is correct and using that alone would be enough?

File: tests/phpunit/tests/media.php (hunk 13 above)
Comment (joemcgill) [reply tests/phpunit/tests/media.php:4156]: I don't mind leaving it as it is, I just want to make sure this is testing exactly what you intend. Currently, this is doing so indirectly by mocking the output of the excerpt and post thumbnail and then asserting that the overall output matches, which leaves the test a bit more fragile as other things could cause side effects that could make this fail (like the example you provided). Generally, I prefer to see unit tests that specifically test exactly what you are trying to confirm.

File: src/wp-includes/formatting.php (hunk 4 above)
Comment (joemcgill) [reply src/wp-includes/formatting.php:3957]: Ah, that's a good point that we're removing it from a specific priority via the default param. If someone moved the priority themselves then this workaround wouldn't help them at all. Happy to leave it as is then, though adding a note to the inline comment above where we're restoring the filter might be nice.

**Review by joemcgill — APPROVED** [review]

File: tests/phpunit/tests/media.php (hunk 10 above)
Comment (felixarntz) [reply tests/phpunit/tests/media.php]: But `wp_increase_content_media_count()` is not covered by this test, even though it's used. This test covers a logic change in `wp_trim_excerpt()`, which doesn't change the output of that function, but unit tests aren't limited to what a function outputs. FWIW, the fact that this function has side effects isn't great, but that's WordPress :)

I agree `the_excerpt` should probably be removed, since indeed the logic change tested here is in `wp_trim_excerpt()`, but not in `the_excerpt()`. I'll update.

File: tests/phpunit/tests/media.php (hunk 13 above)
Comment (felixarntz) [reply tests/phpunit/tests/media.php:4156]: Maybe a good middle ground that still covers all I mentioned would be to ensure the output _contains_ the expected `img` tag without `loading="lazy"`, i.e. ignore comparing the excerpt output?

File: tests/phpunit/tests/media.php (hunk 9 above)
Comment (felixarntz) [reply tests/phpunit/tests/media.php]: Updated in https://github.com/WordPress/wordpress-develop/pull/4439/commits/7f41cb63d2db2572de8829c3614ff8ed051f0247

File: tests/phpunit/tests/media.php (hunk 10 above)
Comment (felixarntz) [reply tests/phpunit/tests/media.php]: Updated in https://github.com/WordPress/wordpress-develop/pull/4439/commits/d613c309bb7aa18089c38571a39da055f6af41e7

File: tests/phpunit/tests/media.php (hunk 11 above)
Comment (felixarntz) [reply tests/phpunit/tests/media.php]: See above, updated in https://github.com/WordPress/wordpress-develop/commit/7f41cb63d2db2572de8829c3614ff8ed051f0247.

File: tests/phpunit/tests/media.php (hunk 13 above)
Comment (felixarntz) [reply tests/phpunit/tests/media.php:4156]: Updated in https://github.com/WordPress/wordpress-develop/pull/4439/commits/0488073761c5597f7f7927c204c35a9f99f951c3.

File: src/wp-includes/formatting.php (hunk 4 above)
Comment (felixarntz) [reply src/wp-includes/formatting.php:3957]: Clarified in a comment in https://github.com/WordPress/wordpress-develop/pull/4439/commits/49a8a52cda65a60d5d2e2df973c3390a4dbd1490.

File: tests/phpunit/tests/media.php [hunk 15]
```diff
@@ -75,6 +75,16 @@ public static function tear_down_after_class() {
 		parent::tear_down_after_class();
 	}
 
+	/**
+	 * Ensures that the static content media count and related filter are reset between tests.
+	 */
+	public function set_up() {
```
Comment (joemcgill) [inline tests/phpunit/tests/media.php:81]: ✨

**Comment by felixarntz** [comment]
Committed in https://core.trac.wordpress.org/changeset/55850
